File: 2024/day1.py
from collections import Counter

from utils import get_file_output

output = get_file_output("day1.txt")

list1 = []
list2 = []
for line in output:
    numbers = line.strip().split(" ")
    list1.append(int(numbers[0]))
    list2.append(int(numbers[-1]))


def part1():
    list1.sort()
    list2.sort()

    total_distance = 0
    i = 0
    while i < len(list1):
        total_distance += abs(list1[i] - list2[i])
        i += 1

    print(total_distance)


def part2():
    counts = Counter(list2)
    i = 0
    total_value = 0
    while i < len(list1):
        value = list1[i]
        new_value = 0
        if value in counts:
            new_value = value * counts[value]

        total_value += new_value
        i += 1

    print(total_value)

if __name__ == '__main__':
    part1()
    part2()

File: 2024/day1.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"slices"
	"strconv"
	"strings"
)

func check(e error) {
	if e != nil {
		panic(e)
	}
}

func getOutput() ([]int, []int) {
	f, err := os.Open("input/day1.txt")

	check(err)
	scanner := bufio.NewScanner(f)
	list1 := make([]int, 0)
	list2 := make([]int, 0)

	for scanner.Scan() {
		line := scanner.Text()
		trimmed_line := strings.Trim(line, " ")
		split_line := strings.Split(trimmed_line, " ")
		last_index := len(split_line) - 1

		first_number, _ := strconv.Atoi(split_line[0])
		second_number, _ := strconv.Atoi(split_line[last_index])

		list1 = append(list1, first_number)
		list2 = append(list2, second_number)
	}

	return list1, list2
}

func partOne() {
	list1, list2 := getOutput()

	slices.Sort(list1)
	slices.Sort(list2)

	total_distance := 0

	i := 0

	for i < len(list1) {
		diff := list1[i] - list2[i]
		total_distance += int(math.Abs(float64(diff)))

		i += 1
	}

	fmt.Println(total_distance)
}

func partTwo() {
	list1, list2 := getOutput()
	counts := map[int]int{}

	for _, num := range list2 {
		counts[num] += 1
	}

	total_value := 0

	for _, num := range list1 {
		count, ok := counts[num]
		new_value := 0
		if ok {
			new_value = num * count
		}

		total_value += new_value
	}

	fmt.Println(total_value)
}

func main() {
	partOne()
	partTwo()
}
